src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,430 @@
+        | None -> f log
+        | Some retryPolicy ->
+            let withLoggingContextWrapping count =
+                let log = if count = 1 then log else log |> prop contextLabel count
+                f log
+            retryPolicy withLoggingContextWrapping
+    let internal (|BlobLen|) (x : Format) = x.Length
+    let (|StrLen|) = function null -> 0 | (x : string) -> x.Length
+
+    /// NB Caveat emptor; this is subject to unlimited change without the major version changing - while the `dotnet-templates` repo will be kept in step, and
+    /// the ChangeLog will mention changes, it's critical to not assume that the presence or nature of these helpers be considered stable
+    module InternalMetrics =
+
+        module Stats =
+            let inline (|Stats|) ({ interval = i } : Measurement) = int64 i.ElapsedMilliseconds
+
+            let (|Read|Write|Resync|Rollup|) = function
+                | Slice (Stats s) -> Read s
+                | Last (Stats s) -> Read s
+                // slices are rolled up into batches so be sure not to double-count
+                | Batch (_, Stats s) -> Rollup s
+                | WriteSuccess (Stats s) -> Write s
+                | WriteConflict (Stats s) -> Resync s
+            type Counter =
+                { mutable count : int64; mutable ms : int64 }
+                static member Create() = { count = 0L; ms = 0L }
+                member x.Ingest(ms) =
+                    System.Threading.Interlocked.Increment(&x.count) |> ignore
+                    System.Threading.Interlocked.Add(&x.ms, ms) |> ignore
+            type LogSink() =
+                static let epoch = System.Diagnostics.Stopwatch.StartNew()
+                static member val Read = Counter.Create() with get, set
+                static member val Write = Counter.Create() with get, set
+                static member val Resync = Counter.Create() with get, set
+                static member Restart() =
+                    LogSink.Read <- Counter.Create()
+                    LogSink.Write <- Counter.Create()
+                    LogSink.Resync <- Counter.Create()
+                    let span = epoch.Elapsed
+                    epoch.Restart()
+                    span
+                interface Serilog.Core.ILogEventSink with
+                    member _.Emit logEvent = logEvent |> function
+                        | MetricEvent (Read stats) -> LogSink.Read.Ingest stats
+                        | MetricEvent (Write stats) -> LogSink.Write.Ingest stats
+                        | MetricEvent (Resync stats) -> LogSink.Resync.Ingest stats
+                        | MetricEvent (Rollup _) -> ()
+                        | _ -> ()
+
+        /// Relies on feeding of metrics from Log through to Stats.LogSink
+        /// Use Stats.LogSink.Restart() to reset the start point (and stats) where relevant
+        let dump (log : ILogger) =
+            let stats =
+              [ "Read", Stats.LogSink.Read
```

Comment:
prob break out ReadL vs Read

Reply:
Counterpoint: 
The SSS metrics don't break out reading forwards and backwards, and since reading the last event is analogous to reading backwards with a max count of 1 I'd argue we should align the stats between SSS and MessageDB.

Reply:
It's more about the higher level operation - i.e. point reads vs scans/queries.
While it may internally be 'just a query', the point is that it's an optimised path - if we end up doing queries for that category, there's a problem.
It's definitely debatable how to break them down though.
(e.g. In DynamoStore, I don't have a single "Append" operation, there's "Append" and "Calve" depending on whether a page split was needed)

In a prometheus sink for Mdb, I'd have ReadL vs Read operations - which lets you split e.g. which reads could be improved by adjusting pagesize, vs those which are optimized point reads.

Reply:
Makes sense to me.

I've added a `ReadL` and modified the tests to expect a `EsAct.ReadLast` followed by a `EsAct.Append` rather than the slice -> batch -> append we see right now.